### ISSUE TITLE
`nc_close` should not require the GIL

### ISSUE THREAD

**Issue by sternj** [Unidata/netcdf4-python#1180]
When running NetCDF4 using [Scalene](https://github.com/plasma-umass/scalene) I ran into an issue where one thread held HDF5's lock without the GIL and was trying to acquire the GIL while another thread was holding the GIL while trying to acquire HDF5's lock. I tracked down what I suspect to be the root of the problem to [the Python declaration of `netcdf4.h`](https://github.com/Unidata/netcdf4-python/blob/master/include/netCDF4.pxi#L324), where the `nc_close` function is declared without a `nogil`. 

I'm trying out a patched version with the `nc_close` function being declared with `nogil` and the offending call to the `nc_close` function being wrapped in a `nogil` block to see if this resolves the issue that I've been seeing, and I can of course PR this, but I wanted to know if there's any specific reason why it was written without the `nogil` in the first place.

**Comment by jswhit** [comment]
Don't know why nc_close is not wrapped 'with nogil' in _netCDF4.pyx.  I don't see why adding that would be a problem - but my understanding of how the GIL works is admittedly limited.

**Comment by sternj** [comment]
I've been digging into this further and found several similar functions which take an `ncid` (and thus HDF5's lock). I can't quite remember where I read this but some of Python's documentation indicates that it is best practice to release the GIL when waiting on a blocking operation.

I actually am seeing a few other functions like this, there are several that are namespaced in that don't have `nogil`, and I'm not sure why any of the functions would need the GIL-- I don't want to put the work into a PR if there's a compelling reason why these do have the GIL. Does anyone know why it was designed this way?

**Comment by dopplershift** [comment]
So long as these functions aren't doing any Python C-API work, releasing the GIL should be fine. Also, releasing the GIL is probably the most (only?) reliable way to deal with potential deadlocks due to competing locks.

It looks like all of the functions in that header are just direct wraps of the netCDF C API, so the GIL should be completely safe to release.

**Comment by sternj** [comment]
I'll have a PR up in the next few days, it seems like it's not quite possible to declare `cdef` variables inside of a `nogil` block (well, the compiler complains very loudly at me), is there a convention that works better than forward-declaring the variable?

**Comment by dopplershift** [comment]
I don't think you can wrap a variable. "nogil" means "release the GIL when calling this function", so it only makes sense to apply to functions/methods.

**Comment by sternj** [comment]
I mean that something like
```cython
with nogil:
   cdef int foo = some_function()
```
whereas
```cython
cdef int foo = -1
with nogil:
   foo = some_function()
  ```
does work

**Comment by dopplershift** [comment]
Oh, weird. Are there places where that's necessary? I would have thought just adding `nogil` to as many function definitions as possible would be enough.

**Comment by sternj** [comment]
This is not quite the case, at least [according to this](https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#declaring-a-function-as-callable-without-the-gil), you need to manually release the GIL I think, unless I'm entirely misreading what `nogil` in a declaration means

**Comment by sternj** [comment]
(as an answer to my question, though, [someone else seems to have established this practice for the repo](https://github.com/Unidata/netcdf4-python/blob/9397b43b65c6b9856c58dc1b448c60b5d2f58ecb/src/netCDF4/_netCDF4.pyx#L1267), which I will follow

**Comment by jswhit** [comment]
This turns out to be a lot more work than I thought - I've been inconsistent in releasing the GIL when calling the C lib.  PR #1180.

**Comment by jswhit** [comment]
The MPI tests are failing for PR #1180

**Comment by jswhit** [comment]
closed by PR #1180
